contrib/src/main/java/com/datatorrent/contrib/enrich/AbstractEnricher.java

```diff
@@ -0,0 +1,319 @@
+
+import org.apache.hadoop.classification.InterfaceStability;
+
+import com.esotericsoftware.kryo.NotNull;
+import com.datatorrent.api.Context;
+import com.datatorrent.api.Operator;
+import com.datatorrent.common.util.BaseOperator;
+import com.datatorrent.lib.db.cache.CacheManager;
+import com.datatorrent.lib.db.cache.CacheStore;
+import com.datatorrent.lib.util.FieldInfo;
+import com.datatorrent.lib.util.FieldInfo.SupportType;
+
+/**
+ * Base class for Enrichment Operator.&nbsp; Subclasses should provide implementation to getKey and convert.
+ * The operator receives a tuple and emits enriched tuple based on includeFields and lookupFields. <br/>
+ * <p>
+ * Properties:<br>
+ * <b>lookupFields</b>: List of comma separated keys for quick searching. Ex: Field1,Field2,Field3<br>
+ * <b>includeFields</b>: List of comma separated fields to be replaced/added to the input tuple. Ex: Field1,Field2,Field3<br>
+ * <b>store</b>: Specify the type of loader for looking data<br>
+ * <br>
+ *
+ * @param <INPUT>  Type of tuples which are received by this operator</T>
+ * @param <OUTPUT> Type of tuples which are emitted by this operator</T>
+ * @displayName Abstract Enrichment Operator
+ * @tags Enrichment
+ */
+@InterfaceStability.Evolving
+public abstract class AbstractEnricher<INPUT, OUTPUT> extends BaseOperator implements Operator.ActivationListener
+{
+  /**
+   * Mandatory parameters for Enricher
+   */
+  @NotNull
+  protected List<String> lookupFields;
+  @NotNull
+  protected List<String> includeFields;
+  @NotNull
+  private BackendLoader store;
+
+  /**
+   * Optional parameters for enricher.
+   */
+  private int cacheExpirationInterval = 24 * 60 * 60 * 1000;
+  private int cacheCleanupInterval = 24 * 60 * 60 * 1000;
+  private int cacheSize = 1024;
+
+  /**
+   * Helper variables.
+   */
+  private transient CacheManager cacheManager;
+  protected transient List<FieldInfo> lookupFieldInfo = new ArrayList<>();
+  protected transient List<FieldInfo> includeFieldInfo = new ArrayList<>();
+
+  /**
+   * This method needs to be called by implementing class for processing a tuple for enrichment.
+   * The method will take th tuple through following stages:
+   * <ol>
+   * <li>Call {@link #getKey(Object)} to retrieve key fields for lookup</li>
+   * <li>Using key fields call cache manager to retrieve for any key that is cached already</li>
+   * <li>If not found in cache, it'll do a lookup in configured backend store</li>
+   * <li>The retrieved fields will be passed to {@link #convert(Object, Object)} method to create the final object</li>
+   * <li>Finally {@link #emitTuple(Object)} is called for emitting the tuple</li>
+   * </ol>
+   *
+   * @param tuple Input tuple that needs to get processed for enrichment.
+   */
+  protected void enrichTuple(INPUT tuple)
+  {
+    Object key = getKey(tuple);
+    if (key != null) {
+      Object result = cacheManager.get(key);
+      OUTPUT out = convert(tuple, result);
+      emitTuple(out);
```

Comment:
Can you rename it to something else as it will confuse with emitTuple of input adapter.
Suggestion: emitEnrichedTuple()